File: December2021/C++/Maximal_Square.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int m=matrix.size(),n=matrix[0].size();
        vector<vector<int>> dp(m+1,vector<int>(n+1,0));
        int res=0;
        for(int i=0;i<m;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(matrix[i][j]=='1')
                {
                    // cout<<i<<" "<<j<<endl;
                    dp[i+1][j+1]=1+min({dp[i][j],dp[i][j+1],dp[i+1][j]});
                    // cout<<"dp val : "<<dp[i+1][j+1]<<endl;
                    res=max((int)res,(int)dp[i+1][j+1]);
                }
            }
        }
        return res*res;
    }
};

File: December2021/MaximalSquare.java
// TC : O(n*m)
// SC. : O(n*m)

class Solution {
	public int maximalSquare(char[][] matrix) {
		if(matrix.length ==0 || matrix[0].length == 0){
			return 0;
		}
		int[][] dp = new int[matrix.length+1][matrix[0].length+1];
		int ans = 0; // sqaure length
		for(int i=1;i<=matrix.length;i++){

			for(int j=1;j<=matrix[0].length;j++) {
				if(matrix[i-1][j-1] == '1') {
					int min = Integer.MAX_VALUE;
					min = Math.min(dp[i-1][j], min);
					min = Math.min(dp[i][j-1], min);
					min = Math.min(dp[i-1][j-1], min);

					dp[i][j] = min+1;
					ans = Math.max(dp[i][j], ans);
				}
			}
		}

//          for(int i=0;i<=matrix.length;i++){

//             for(int j=0;j<=matrix[0].length;j++) {
//                     System.out.print(dp[i][j] + " ");
//                 }
//              System.out.println();
//             }

		return ans * ans;

	}
}
